feat: Replace Generic Errors with Suspicious Transaction Errors for UTXO/EthLike Coins

danielpeng1 opened the pull request:
Replace generic Error throws with `TxIntentMismatchError` in `verifyTransaction` methods for Bitcoin and other UTXO & EthLike coins to provide more specific error handling for suspicious transaction detection.
- Updated UTXO `verifyTransaction` methods
- Updated EthLike `verifyTransaction` methods

Open to comments relating to copy or if I've missed any generic error.

Ticket: WP-6189

danielpeng1 committed:
feat: replace generic Error with TxIntentMismatchError

Ticket: WP-6189

yarikpopov164 reviewed (dismissed):
Not sure if this task can be done before the #7148 is merged

danielpeng1 commented on modules/sdk-core/src/bitgo/errors.ts:
```diff
@@ -157,6 +157,12 @@ export class InvalidTransactionError extends BitGoJsError {
   }
 }
 
+export class SuspiciousTransactionError extends BitGoJsError {
```
Yeah, I think the name was changed and needs to be updated on the BTC ticket. Let me pull Mohammad in for a review.

OttoAllmendinger commented on modules/abstract-utxo/src/transaction/fixedScript/verifyTransaction.ts:
```diff
@@ -25,19 +25,41 @@ function getPayGoLimit(allowPaygoOutput?: boolean): number {
 
+  // Helper to throw TxIntentMismatchError with consistent context
+  const throwTxMismatch = (message: string): never => {
+    throw new TxIntentMismatchError(message, reqId || '', [txParams], txPrebuild.txHex || '');
```
Why pass empty strings here? Change the TxIntentMismatchError to accept `undefined` instead

danielpeng1 committed:
refactor: update TxIntentMismatchError to use undefined

Ticket: WP-6189

OttoAllmendinger commented on modules/abstract-utxo/src/transaction/fixedScript/verifyTransaction.ts:
```diff
@@ -75,7 +97,7 @@ export async function verifyTransaction<TNumber extends bigint | number>(
     const isBitgoKeySignatureValid = verify(keychains.bitgo, keySignatures.bitgoPub);
     if (!isBackupKeySignatureValid || !isBitgoKeySignatureValid) {
-      throw new Error('secondary public key signatures invalid');
+      throwTxMismatch('secondary public key signatures invalid');
```
not an intent mismatch

OttoAllmendinger commented on modules/abstract-utxo/src/transaction/fixedScript/verifyTransaction.ts:
```diff
@@ -86,11 +108,11 @@ export async function verifyTransaction<TNumber extends bigint | number>(
     const customChangeKeySignaturesVerified = verifyCustomChangeKeySignatures(parsedTransaction, keychains.user);
     if (!customChangeKeySignaturesVerified) {
-      throw new Error(
+      throwTxMismatch(
```
not an intent mismatch

OttoAllmendinger commented on modules/abstract-utxo/src/transaction/fixedScript/verifyTransaction.ts:
```diff
@@ -100,7 +122,7 @@ export async function verifyTransaction<TNumber extends bigint | number>(
   if (missingOutputs.length !== 0) {
     // there are some outputs in the recipients list that have not made it into the actual transaction
-    throw new Error('expected outputs missing in transaction prebuild');
+    throwTxMismatch('expected outputs missing in transaction prebuild');
```
this is plausibly an actual intent mismatch

lcovar replied:
agree

OttoAllmendinger commented on modules/abstract-utxo/src/transaction/fixedScript/verifyTransaction.ts:
```diff
@@ -163,7 +185,7 @@ export async function verifyTransaction<TNumber extends bigint | number>(
 
   if (fee < 0) {
-    throw new Error(
+    throwTxMismatch(
```
not an intent mismatch

lcovar committed:
feat: improve transaction verification error handling

Replace custom validation errors with SDK's TxIntentMismatchRecipientError to
provide more detailed information about mismatched transaction outputs. This
enhances error reporting by including specific details about missing or
unexpected outputs, helping users understand what differs between their intent
and the actual transaction.

Co-authored-by: llm-git <llm-git@ttll.de>

Ticket: BTC-2579

TICKET: WP-6189

OttoAllmendinger commented on modules/abstract-utxo/src/transaction/descriptor/verifyTransaction.ts:
```diff
@@ -1,50 +1,54 @@
-    super(`unexpected implicit external outputs (count=${implicitExternalOutputs.length})`);
-  }
-}
-
```
You are removing useful structure from the errors here.

Instead of rewriting this file, consider catching the error on in `abstractUtxoCoin.ts` and re-throwing it wrapped in a `TxIntentMismatchError`

If it is important that the outer error is of class `TxIntentMismatchRecipientError`, please wrap the structured error inside. Consider passing it as cause: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause

danielpeng1 committed:
refactor: enhance eth-like tx verification with recipient mismatch errors

Ticket: WP-6189

OttoAllmendinger commented on modules/abstract-utxo/src/transaction/descriptor/verifyTransaction.ts:
can you move all the changes in this file to AbstractUtxoCoin? Do the catch + conversion there?

danielpeng1 committed:
feat: refactoring error handling for transaction verification

Ticket: WP-6189
